File: 0539-minimum-time-difference/0539-minimum-time-difference.cpp
class Solution {
public:
    int findMinDifference(vector<string>& timePoints) {
        int mnMins=INT_MAX,n=timePoints.size();
        sort(timePoints.begin(),timePoints.end());
        for(int i=0;i<n;i++){
            int mins=abs(findDiff(timePoints[(i-1+n)%n],timePoints[i]));
            int minDiff=min(mins,1440-mins);
            mnMins=min(minDiff,mnMins);
        }
        return mnMins;
    }
    int findDiff(string s1,string s2){
        int h1=stoi(s1.substr(0,2));
        int h2=stoi(s2.substr(0,2));
        int m1=stoi(s1.substr(3,2));
        int m2=stoi(s2.substr(3,2));
        return (h2-h1)*60 + (m2-m1);
    }
    
};

File: 0539-minimum-time-difference/0539-minimum-time-difference.java
class Solution {
    public int findMinDifference(List<String> timePoints) {
        boolean[] bo=new boolean[1440];
        for(String s:timePoints){
            String[] t=s.split(":");
            int h=Integer.parseInt(t[0]);
            int m=Integer.parseInt(t[1]);
            if(bo[h*60+m])return 0;
            bo[h*60+m]=true;
        }
        int mnMins=Integer.MAX_VALUE,prev=Integer.MAX_VALUE,last=Integer.MIN_VALUE,ptr=0;
        for(int i=0;i<1440;i++){
            if(bo[i]){
                if(prev!=Integer.MAX_VALUE){
                    mnMins=Math.min(mnMins,i-ptr);
                }
                prev=Math.min(prev,i);
                last=Math.max(last,i);
                ptr=i;
            }
        }
        mnMins=Math.min(mnMins,1440-(last-prev));
        return mnMins;
    }
}